nomad/state/state_store.go

```diff
@@ -437,6 +449,86 @@ func (s *StateStore) DeleteJob(index uint64, jobID string) error {
 	return nil
 }
 
+// deleteJobVersions deletes all versions of the given job.
+func (s *StateStore) deleteJobVersions(index uint64, job *structs.Job, txn *memdb.Txn) error {
+	iter, err := txn.Get("job_versions", "id_prefix", job.ID)
+	if err != nil {
+		return err
+	}
+
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+
+		// Ensure the ID is an exact match
+		j := raw.(*structs.Job)
+		if j.ID != job.ID {
+			continue
+		}
+
+		if _, err = txn.DeleteAll("job_versions", "id", job.ID, job.Version); err != nil {
+			return fmt.Errorf("deleing job versions failed: %v", err)
+		}
+	}
+
+	if err := txn.Insert("index", &IndexEntry{"job_summary", index}); err != nil {
+		return fmt.Errorf("index update failed: %v", err)
+	}
+
+	return nil
+}
+
+// upsertJobVersion inserts a job into its historic version table and limits the
+// number of job versions that are tracked.
+func (s *StateStore) upsertJobVersion(index uint64, job *structs.Job, txn *memdb.Txn) error {
+	// Insert the job
+	if err := txn.Insert("job_versions", job); err != nil {
+		return fmt.Errorf("failed to insert job into job_versions table: %v", err)
+	}
+
+	if err := txn.Insert("index", &IndexEntry{"job_versions", index}); err != nil {
```

Comment:
Yeah we still need this. It is used to return the index to wait on when the blocking query is against a nonexistent object. Consul does the same but with slightly different semantics (getter methods on state store return the index rather than the endpoints looking it up.)